python/plugins/processing/algs/qgis/Dissolve.py

```diff
@@ -49,17 +50,24 @@ class Dissolve(GeoAlgorithm):
 
     def processAlgorithm(self, progress):
         useField = not self.getParameterValue(Dissolve.DISSOLVE_ALL)
-        fieldname = self.getParameterValue(Dissolve.FIELD)
         vlayerA = dataobjects.getObjectFromUri(
             self.getParameterValue(Dissolve.INPUT))
-        field = vlayerA.fieldNameIndex(fieldname)
         vproviderA = vlayerA.dataProvider()
-        fields = vproviderA.fields()
+        fields = QgsFields()
+        if useField:
+            fieldname = self.getParameterValue(Dissolve.FIELD)
+            fieldIdx = vlayerA.fieldNameIndex(fieldname)
```

Comment:
It's dangerous to request the index from the layer and then use it to get a field from the provider.
I would try to avoid direct access to the provider wherever possible. `field = vlayerA.field(fieldname)`should be sufficient.